### ISSUE TITLE
Configurable dimension type

### ISSUE THREAD

**Issue by psiha** [jfalcou/kiwaku#6]
For example 16bit ints are quite enough for our use case and an array of 4 shorts (for 4d arrays - typical "tensors" used in ML) fits into a single 64bit register - so a fully dynamic view can be passed by value with only two 64bit registers (as opposed to five if you use size_ts)...

**Comment by jfalcou** [comment]
Good point. We ponderef about this but wzrent sure of the utility. Glad to have confirmation.

We'll see where we can pass this information

**Comment by jfalcou** [comment]
Updated API being developed in #13 :

```c++
auto s1  = kwk::extent(1,2,3); // [1 2 3] as a shape stored in ptrdiff_t
auto s2  = kwk::extent(1,kwk::fixed<2>,3); // [1 2 3] as a shape stored in ptrdiff_t but 2 is compile-time info
auto s2b = kwk::extent(1,2_c,3); // Same but use the _c literal suffix
auto s3  = kwk::extent_of<short>(1,2,3); // [1 2 3] as a shape stored as short
```

The old `shape<extent()[4]>`is now `shape<shape_()[4]>` but only generates ptrdiff_t based shape storage.
I wen against deducing the storage type for shape based on the type of w/e it is passed as argument cause it would lead to subtle bugs and writing short(4) everywhere seemed bad.

`extent_of<T>` is the best I found atm. If you have a better naming idea, I'm open to hearing it.

Sample codegen:

```c++
auto f(short n) {  return kwk::extent_of<short>(n,2*n); }
```

becomes:

```c++
f(short):                                  # @f(short)
        movzwl  %di, %eax
        shll    $17, %edi
        orl     %edi, %eax
        retq
```

**Comment by jfalcou** [comment]
It's not a move away, extent is just a function that computes the correct shape_ types without users having to deal with it.

My first idea was to have:

```c++
shape< shape_<short>()[4] > x;
shape< shape_()[4] > y;
```

but shape_ being a template variable, it can't exist in both ways.

The function is suffering from the following shortcomings. Say we have just extent(Dims...) and extent<T>(Dims...).
When you pass a single value, the call is ambiguous on some compilers. But this is something I can probably fix, thus not needing two names indeed.

Good catch on the common_type, something like using `common_type<T...>::type` is maybe a good idea.

**Comment by jfalcou** [comment]
Yeah, OK, fixed the naming thing. I had fatty fingers and wrote crap.

**Comment by psiha** [comment]
> It's not a move away, extent is just a function that computes the correct shape_ types without users having to deal with it.

hm so let's see if i follow it correctly now:
* extent can both be used as a variadic function for creating shapes - extent( 1, 2, 3, 4)
* _and_ as the object providing the operator() & operator[] funkiness behaviour/syntax - extent[ 1 ]()[ 4 ]
?

And now you have just
> `shape<extent()[4]>`is now `shape<shape_()[4]>`

renamed extent to shape_?
If so - why? 😂

**Comment by jfalcou** [comment]
I need multiple things:
 - shape : the type for the dimension layout. It's a type that is defined with a shape layout
 - the shape layout is the funky ()[]-supporting object. It was named extent, now it is named shape_ for consistency.
 - extent is now the function that allow for fine-grain control on constructing shape without dealing with shape_[]()()
 
 Cool thing: I'm making sure we write the doc alongside so when #13 get merged, exampels will be in.

**Comment by jfalcou** [comment]
Also note, than before any v0.1 tagged release, we're tryign to find the better naming/setup before freezing the API.

**Comment by jfalcou** [comment]
Yeah I know. Again I may find a better name

**Comment by jfalcou** [comment]
OK, mulled over the nameing scheme. We're back to
shape/extent[]()/ and of_size for the functional generator. No more ugly shape_

**Comment by psiha** [comment]
so how do i do this in the end? 😁 
extent does not support the underlying-type template argument? i have to use detail::shaper or?

**Comment by jfalcou** [comment]
https://github.com/jfalcou/kiwaku/blob/main/test/container/shape/mixed.cpp

For how to make of_size change types.

Do you need an helper for the type or ?

**Comment by jfalcou** [comment]
Ok I see what is missing.

**Comment by psiha** [comment]
tried shape< extent< short >()()()() > but that does not work obviously - i need to use of_size for that?

**Comment by jfalcou** [comment]
Ok so here are the options once the new PR lands:

```c++
#include<kwk/container/shape.hpp>

// Shape descriptors with runtime sizes
kwk::shape< kwk::extent()()           > d1(6, 8); // use ptrdiff_t
kwk::shape< kwk::extent_of<short>()() > s1(6, 8); // use short

// Shape with of_size
auto p1 = kwk::of_size(4,4); // use int internally (because 4 is int)
auto q1 = kwk::of_size<short>(4,4); // use short

short n = 8, m = 7;
auto r1 = kwk::of_size(n,m); // use short
```

**Comment by jfalcou** [comment]
Can you show me your call exactly ?

**Comment by psiha** [comment]
> Can you show me your call exactly ?

bit complicated

```
    template < auto ... targetDims >
    constexpr operator Dimensions< targetDims... >() const noexcept requires( equivalent( Dimensions< targetDims... >{}, Dimensions< dims... >{} ) )
    {
        return {{ kwk::of_size< shape_t >( Detail::getDimForKWKCopyConstruction< targetDims >( *this )... ) }};
    }
```
    
(implementing/experimenting layout support/typed dimensions locally😊)

**Comment by jfalcou** [comment]
> kiwaku\include\kwk/container/shape.hpp(544,18): message : candidate function [with SizeType = unsigned short, Ds = <unsigned short, unsigned short, std::integral_constant<unsigned char, '\x01'>>]
kiwaku\include\kwk/container/shape.hpp(570,43): message : candidate function [with Ds = <unsigned short, unsigned short, std::integral_constant<unsigned char, '\x01'>>]

so this means you do a `of_size<unsigned short>(us,1_c)` ?

**Comment by jfalcou** [comment]
OK reproduced.

**Comment by jfalcou** [comment]
OK, I guess I was a bit optimistic. There is some ambiguity in the current code.
Are with ok with having a `of_size` and a `of_size_with<T>` ?

**Comment by psiha** [comment]
> OK, I guess I was a bit optimistic. There is some ambiguity in the current code. Are with ok with having a `of_size` and a `of_size_with<T>` ?

oh well...let's go with that for starters :)

**Comment by jfalcou** [comment]
it can also be a parameters if you want

`of_size(type<T>, ds...)`

**Comment by jfalcou** [comment]
So alternative works:

```
of_size(as<short>, v1, v2);
```

which one do you prefer ?

**Comment by psiha** [comment]
> which one do you prefer ?

of_size_with<T> (at least this should compile faster than additional type wrappers:)

**Comment by jfalcou** [comment]
:+1: Shipping away in a few mn

**Comment by jfalcou** [comment]
See: https://github.com/jfalcou/kiwaku/pull/30

**Comment by jfalcou** [comment]
#30  has been merged. If you can check on your side it's OK.
I added the use case to the unit tests though.

**Comment by psiha** [comment]
this works now 👍
now waiting for https://github.com/jfalcou/kiwaku/issues/26 😁
